docs/optlib.rst

```diff
@@ -181,269 +206,413 @@ Tips about writing option file
+single literal character control codes before passing the pattern to glibc.
+
+Another detail to keep in mind is how the regex engine treats newlines.
+Universal-ctags compiles the regular expressions in the ``--regex-<LANG>`` and
+``--mline-regex-<LANG>`` options with REG_NEWLINE set. What that means is documented
+in the
+`POSIX spec <http://pubs.opengroup.org/onlinepubs/009695399/functions/regcomp.html>`_.
+One obvious effect is that the regex special dot any-character ``.`` does not match
+newline characters, the ``^`` anchor *does* match right after a newline, and
+the ``$`` anchor matches right before a newline. A more subtle issue is this text from the
+`Regular Expressions chapter <http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap09.html>`_:
+"the use of literal <newline>s or any escape sequence equivalent produces undefined
+results". What that means is using a regex pattern with ``[^\n]+`` is invalid,
+and indeed in glibc produces very odd results. **Never** use ``\n`` in patterns
+for ``--regex-<LANG>``, and never use them in non-matching bracket expressions
+for ``--mline-regex-<LANG>`` patterns. For the experimental ``--_mtable-regex-<LANG>``
+you can safely use ``\n`` because that regex is not compiled with REG_NEWLINE.
+
+You should always test your regex patterns against test files with strings that
+do and do not match. Pay particular emphasis to when it should *not* match, and
+how *much* it matches when it should. A common error is forgetting that a
+POSIX.2 ERE engine is always greedy; the `*` and `+` quantifiers match
+as much as possible, before backtracking from the end of their match.
+
+For example this pattern::
+
+	foo.*bar
+
+Will match this **entire** string, not just the first part::
+
+	foobar, bar, and even more bar
+
+
+Regex option argument flags
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Even if more flags are added like ``x``, ``y``, ``z``,..., users
-may not utilize them well because it is difficult to memorize them. In
-addition, if many "option libraries" are contributed, we have to
-maintain them.
+Many regex-based options described in this document support additonal arguments
+in the form of long flags. Long flags are specified with surrounding ``{`` and
+``}``.
 
-For both users and developers the variety of short flags are just
-nightmares.
+The general format and placement is as follows::
 
-So universal-ctags now includes an API for defining long flags, which can be
-used as aliases for short flags. The long flags requires more typing
-but are more readable.
+	--regex-<LANG>=<PATTERN>/<NAME>/[<KIND>/]LONGFLAGS
 
-Here is the mapping between the standard short flag names and long flag names:
+Some examples:
 
-=========== ===========
-short flag  long flag
-=========== ===========
-b           basic
-e           extend
-i           icase
-=========== ===========
+.. code-block:: perl
```

Comment:
Should we use perl here?
For highlighting the regex patterns?

Reply:
It shows fairly nicely in the generated HTML.